### ISSUE TITLE
Idea: shell-style < operator to read file contents

### ISSUE THREAD

**Issue by leoj3n** [lorenwest/node-config#531]
I know the wiki suggests using `.js` and `fs.readFileSync`:

https://github.com/lorenwest/node-config/wiki/Special-features-for-JavaScript-configuration-files#including-other-files

But often I just need to quickly read a key/cert file and found myself wishing I could just do:

```json
{
  "key": "< ../ssl/key.pem",
  "cert": "< ../ssl/cert.pem"
}
```

In my `development.json` for example. This seemed so natural to me I thought I might suggest it here.

Is file resolve already supported in this package? I come from using [`@feathersjs/configuration`](https://github.com/feathersjs/feathers/tree/master/packages/configuration) that does [this](https://github.com/feathersjs/feathers/blob/cd4b64419cea4f6c0a89a026a3c0a8875eee4996/packages/configuration/lib/index.js#L28-L34) so maybe this makes more sense to suggest there.

**Comment by leoj3n** [comment]
@lorenwest Cool, that would save me from having to do https://github.com/lorenwest/node-config/issues/253#issuecomment-457035455 in all my configs.

**Comment by iMoses** [comment]
@markstos can be considered an alternative parser which can be defined with `Parser`.
Feels to me that at this point if we would like to create a template language it's best to be a part of a sibling package and not this one.

**Comment by leoj3n** [comment]
HCL (Hashicorp Markup Lang) might be relevant to this discussion as well?

On July 11, 2019 6:07:27 PM CDT, iMoses <notifications@github.com> wrote:
>@markstos can be considered an alternative parser which can be defined
>with `Parser`.
>Feels to me that at this point if we would like to create a template
>language it's best to be a part of a sibling package and not this one.
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/lorenwest/node-config/issues/531#issuecomment-510685555

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by iMoses** [comment]
@leoj3n you can simply add HCL support if you have a parser to translate HCL to JSON, or you're willing to create one.

For example, use can use [hcl-to-json](https://www.npmjs.com/package/hcl-to-json)
```
const hcltojson = require('hcltojson');
const Parser = require('config/parser');
Parser.setParser('hcl', (filename, content) => hcltojson(content));
module.exports = Parser;
```

**Comment by markstos** [comment]
Closing in favor of @iMoses suggestion.

**Comment by leoj3n** [comment]
Did not entirely follow the conversation between you two here. Have you guys taken over for lorenwest and now deciding where to draw the line between configuration versus templating language? What I was thinking was HCL seems to have invented the wheel as far as striking a good balance?

After investigating node config I'm thinking it is not striking such a good balance and just adding complexity overall where a simpler config scheme would be more straightforward.

On July 12, 2019 9:01:51 AM CDT, iMoses <notifications@github.com> wrote:
>@leoj3n you can simply add HCL support if you have a parser to
>translate HCL to JSON, or you're willing to create one.
>
>For example, use can use
>[hcl-to-json](https://www.npmjs.com/package/hcl-to-json)
>```
>const hcltojson = require('hcltojson');
>const Parser = require('config/parser');
>Parser.setParser('hcl', (filename, content) => hcltojson(content));
>module.exports = Parser;
>```
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/lorenwest/node-config/issues/531#issuecomment-510897865

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by iMoses** [comment]
@leoj3n Only speaking for myself here...

I believe `node-config` shouldn't be opinionated in order to provide a generic baseline for simple and complex configuration systems and needs. 

Anything we decide in regard to templates will by default be opinionated. What you view as a simple feature I also view as a potential bug for someone who requires the same syntax for their internal usage. That being said, there's currently an option to bypass `config` default parser and provide it with an alternative one, which can handle templates among other things. 

I think that if we want to create our own config template solution we should do so in a separate package. Feel free to take the initiative, publish your solution so that other users will benefit from it and have a chance to grow it as a community, that way you'll also see how many people agree with your template solution and pick it as their standard. If it becomes standard enough we can add it to one of the "soft-dependencies" as another predefined extension.

But I truly believe that if `node-config` claim to parse `json` files, I want them parsed as `json` and nothing else, especially when given the option to change the default parser if desired.

**Comment by leoj3n** [comment]
Ok think I was confused by the sudden convo on this issue. I havent used HCL outside of server config for consul and vault but think it solves some of the problems discussed here by making opinionated decisions.

I was thinking maybe worth refefencing for making design decisions here. Hope it helps if you had not checked out HCL before, as I believe they had to make similar design decisions regarding language versus config.

Lastly to clarify I found the internal node config code itself to be messy, and trying to do too many things, something it sounds you are working on compartmentalizing. More opinion is probably needed here?

Will play with HCL and Parser another later date, haven't looked into all this in detail yet.

On July 12, 2019 12:07:27 PM CDT, iMoses <notifications@github.com> wrote:
>@leoj3n Only speaking for myself here...
>
>I believe `node-config` shouldn't be opinionated in order to provide a
>generic baseline for simple and complex configuration systems and
>needs. 
>
>Anything we decide in regard to templates will by default be
>opinionated. What you view as a simple feature I also view as a
>potential bug for someone who requires the same syntax for their
>internal usage. That being said, there's currently an option to bypass
>`config` default parser and provide it with an alternative one, which
>can handle templates among other things. 
>
>I think that if we want to create our own config template solution we
>should do so in a separate package. Feel free to take the initiative,
>publish your solution so that other users will benefit from it and have
>a chance to grow it as a community, that way you'll also see how many
>people agree with your template solution and pick it as their standard.
>If it becomes standard enough we can add it to one of the
>"soft-dependencies" as another predefined extension.
>
>But I truly believe that if `node-config` claim to parse `json` files,
>I want them parsed as `json` and nothing else, especially when given
>the option to change the default parser if desired.
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/lorenwest/node-config/issues/531#issuecomment-510961704

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by markstos** [comment]
@leoj3n The custom parser functionality is documented here: 

https://github.com/lorenwest/node-config/wiki/Modifying-the-Configuration-Files-Parser

It was just released in node-config 3.2.0.
